src/general/world_effects/patch_events/GlobalGlaciationEvent.cs

```diff
@@ -0,0 +1,258 @@
+﻿using System.Collections.Generic;
+using Godot;
+using Newtonsoft.Json;
+using Xoshiro.PRNG64;
+
+[JSONDynamicTypeAllowed]
+public class GlobalGlaciationEvent : IWorldEffect
+{
+    private const string TemplateBiomeForIceChunks = "ice_shelf";
+    private const string Prefix = "globalGlaciation_";
+    private const string Background = "iceshelf";
+
+    private static readonly string[] IceChunksConfigurations =
+        ["iceShard", "iceChunkSmall", "iceChunkBig", "iceSnowflake"];
+
+    [JsonProperty]
+    private readonly XoShiRo256starstar random;
+
+    [JsonProperty]
+    private readonly List<int> modifiedPatchesIds = new();
+
+    [JsonProperty]
+    private bool hasEventAlreadyHappened;
+
+    /// <summary>
+    ///   Tells how many generations the event will last. "-1" means that it hasn't started at all.
+    ///   "0" means it has finished, and it won't happen again
+    /// </summary>
+    [JsonProperty]
+    private int generationsLeft = -1;
+
+    [JsonProperty]
+    private int eventDuration;
+
+    [JsonProperty]
+    private GameWorld targetWorld;
+
+    public GlobalGlaciationEvent(GameWorld targetWorld, long randomSeed)
+    {
+        this.targetWorld = targetWorld;
+        random = new XoShiRo256starstar(randomSeed);
+    }
+
+    [JsonConstructor]
+    public GlobalGlaciationEvent(GameWorld targetWorld, XoShiRo256starstar random)
+    {
+        this.targetWorld = targetWorld;
+        this.random = random;
+    }
+
+    public void OnRegisterToWorld()
+    {
+    }
+
+    public void OnTimePassed(double elapsed, double totalTimePassed)
+    {
+        if (generationsLeft > 0)
+            generationsLeft -= 1;
+
+        if (generationsLeft > 0)
+            MarkPatches(totalTimePassed);
+
+        if (hasEventAlreadyHappened)
+            return;
+
+        if (generationsLeft == -1)
+        {
+            TryToTriggerEvent(totalTimePassed);
+        }
+        else if (generationsLeft == 0)
+        {
+            FinishEvent();
+        }
+    }
+
+    private void MarkPatches(double totalTimePassed)
+    {
+        foreach (var patch in targetWorld.Map.Patches.Values)
+        {
+            if (patch.Depth[0] == 0)
```

Comment:
This is not consistent with the check in TryToTriggerEvent? Is there a reason? I think having a check method if something qualifies as a surface patch would be better to make into a helper method to ensure one place doesn't check things inconsistently.